### PR TITLE
Added capablity to customize branch protection settings

### PR DESCRIPTION
**Pull request by avraf**
@eyalstoler 
I need to add testing,
but this would be the idea before refactoring.

### CODE REVIEW

**Comment by eyalstoler** [comment]
Looks good :+1:

**Comment by avraf** [comment]
I tried adding the following test
```python
    def test_should_remove_branch_protection(self):
        self.repo.protect_branch("master", True, "everyone", ["test"])
        branch = self.repo.get_protected_branch("master")
        self.assertTrue(branch.protected)
        self.repo.remove_branch_protection('master')
        branch = self.repo.get_protected_branch("master")
        self.assertFalse(branch.protected)
```
It follows their convention:
```python
    def testRemoveBranchProtection(self):
        self.repo.protect_branch("master", False)
        branch = self.repo.get_protected_branch("master")
        self.assertFalse(branch.protected)
        self.assertEqual(branch.enforcement_level, "off")
        self.assertEqual(branch.contexts, [])
```
I get an IO error.
The framework class causes requests to read from a file based on test name.
Please advise.
Otherwise I want to send the code to them and have them direct me.

**Comment by avraf** [comment]
@eyalstoler

**Comment by avraf** [comment]
@eyalstoler,
I would really appreciate your guidance in advancing on this pr.
I wrote tests, which I feel don't really check anything.

**Comment by avraf** [comment]
@eyalstoler 
Just so you know this also works
```python
import github


def remove_branch_protection(self, branch):
    """
    :calls: `DELETE /repos/:owner/:repo/branches/:branch/protection <https://developer.github.com/v3/repos/branches/#remove-branch-protection>`_
    """
    return self._requester.requestJsonAndCheck(
        'DELETE',
        "{base_url}/branches/{branch}/protection".format(base_url=self.url,
                                                            branch=branch),
        headers={'Accept': 'application/vnd.github.loki-preview+json'}
    )

def set_branch_protection(self, branch, require_review=None, restrictions=None, **required_status_checks):
    """
    :calls: `PUT /repos/:owner/:repo/branches/:branch/protection <https://developer.github.com/v3/repos/branches/#update-branch-protection>`_
    :param branch: string
    :param require_review: None or dict with key include_admins set to boolean
    :param restrictions: null or dict with lists as values and keys users and teams
    :param include_admins: boolean
    :
    :param required_status_checks: dict with keys include_admins, strict, and contexts
    """
    protection_url = "{base_url}/branches/{branch}/protection".format(
                                                                base_url=self.url,
                                                                branch=branch)

    if required_status_checks:
        required_checks_url = '{}/required_status_checks'.format(protection_url)
        required_status_checks = dict(required_status_checks,
                                        url=required_checks_url,
                                        contexts_url=required_checks_url+'/contexts')
    else:
        required_status_checks = None
    
    return self._requester.requestJsonAndCheck(
        'PUT',
        protection_url,
        input={
            'required_status_checks': required_status_checks,
            'required_pull_request_reviews': require_review,
            'restrictions' : restrictions
        },
        headers={'Accept': 'application/vnd.github.loki-preview+json'}
    )

github.Repository.Repository.remove_branch_protection = remove_branch_protection
github.Repository.Repository.set_branch_protection = set_branch_protection

k = kgithub
```

File: github/tests/ReplayData/Repository.test_should_add_basic_branch_protection.txt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+200
+[('content-length', '3599'), ('vary', 'Accept, Authorization, Cookie, X-GitHub-OTP, Accept-Encoding'), ('x-served-by', '13d09b732ebe76f892093130dc088652'), ('x-xss-protection', '1; mode=block'), ('x-content-type-options', 'nosniff'), ('etag', '"23247e636875225f1e2d6b49e3ed8fb5"'), ('access-control-allow-credentials', 'true'), ('status', '200 OK'), ('x-ratelimit-remaining', '4997'), ('x-github-media-type', 'github.loki-preview; format=json'), ('access-control-expose-headers', 'ETag, Link, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval'), ('x-github-request-id', '417D879D:14B62:29334564:568D2D11'), ('cache-control', 'private, max-age=60, s-maxage=60'), ('date', 'Wed, 06 Jan 2016 15:04:49 GMT'), ('access-control-allow-origin', '*'), ('content-security-policy', "default-src 'none'"), ('strict-transport-security', 'max-age=31536000; includeSubdomains; preload'), ('server', 'GitHub.com'), ('x-ratelimit-limit', '5000'), ('x-frame-options', 'deny'), ('content-type', 'application/json; charset=utf-8'), ('x-ratelimit-reset', '1452096288')]
```
Comment (eyalstoler) [inline github/tests/ReplayData/Repository.test_should_add_basic_branch_protection.txt]: Change all text below to `[]`

**Comment by avraf** [comment]
@eyalstoler 
I believe I did enough testing.
I also refactored a lot.
We probably could change some variable names.
Everything has default values and now you only have to send one dict as an argument

File: github/Repository.py [hunk 2]
```diff
@@ -2234,6 +2234,51 @@ def protect_branch(self, branch, enabled, enforcement_level=github.GithubObject.
...
+    def set_branch_protection(self, branch, require_review_all=False, require_review=None, restrictions=None, **required_status_checks):
+        """
+        :calls: `PUT /repos/:owner/:repo/branches/:branch/protection <https://developer.github.com/v3/repos/branches/#update-branch-protection>`_
+        :param branch: string
+        :param require_review_all: boolean
+        :param require_review: boolean
+        :param restrictions: dict with lists as values and keys 'users' and 'teams'
+        :param include_admins: boolean
+        :param strict: boolean
+        :param contexts: list
```
Comment (eyalstoler) [inline github/Repository.py]: Add documentation about the default required checks

**Pull request merged**